src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +19,61 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return BATTERY_AMP_HOUR_CAPACITY-AmpHours;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (Current < 0)
+    {
+        return false;
+    }
+    else
+    {
+        return true;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double timeRemaining = (BATTERY_AMP_HOUR_CAPACITY-AmpHours)/inCurrent;
+    timeRemaining = qAbs(timeRemaining)/CONVERT_TO_MILLISECONDS;
+    return QTime(0, 0, 0, 0).addMSecs(timeRemaining);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+
+    QTime currentTime = batteryData.time;
+    changeTime = abs(prevTime.msecsTo(currentTime));
+    changeTime = changeTime * CONVERT_TO_MILLISECONDS;
+    prevTime = currentTime;
+    if (Current > 0 || Current < 0 )
+    {
+        inCurrent = Current;
+    }
+    Current = batteryData.current;
+
+    if (inCurrent == 0)
+    {
+        AvgCurrent = 0;
+    }
+    else
+    {
+        AvgCurrent = (inCurrent + Current)/2;
+    }
+
+    double inVoltage = batteryData.voltage;
+    double Voltage = batteryData.voltage;
+    double AmpChange = (AvgCurrent * changeTime);
+
+    AmpHours += AmpChange;
+
+    int counter;
+    counter++;
```

Comment:
When you initialize `int counter;` here, counter becomes a garbage value. So the `counter++` doesn't give you a value that you know what it is either, when you use it in `SumCurrent/counter` later, counter is a random value that you don't know what it is. Counter should be initialized in the constructor as a private member if you want to carry the value between function calls.